[PM-145] add phishing protection badge

willmartian opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR adds a green badge to the browser extension's icon when the user is on a trusted web vault (either Bitwarden.com or another preconfigured site).

If the badge would show a cipher count, it shows the cipher count with a green background. Otherwise, it shows a green badge with a checkmark. 

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **apps/browser/src/listeners/update-badge.ts:** Added a new method `UpdateBadge.setTrustedSite` that is called at the end of every `run`.

## Screenshots

<img width="585" alt="image" src="https://user-images.githubusercontent.com/17113462/224612475-d2bf8dd6-81e5-47c5-ad11-5ceed67028fe.png">

<img width="429" alt="Screenshot 2023-08-26 at 10 21 33 AM" src="https://github.com/bitwarden/clients/assets/17113462/185e8d8f-c362-4d6f-8710-267ae5d39889">

https://github.com/bitwarden/clients/assets/17113462/b6c459cd-9af5-4a33-aaca-63f17a6f8888

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

willmartian commented:
My first time working with the browser extension! Please ensure I didn't do anything egregious. :)

MGibson1 commented:
I'm hesitant to give up my one spot for text in these badges. I do like the idea to offer some visual indication that you're on the trusted site -- though, I would like it to be limited to _vault_.bitwarden.com (or the exact vault url for self host).

@bitwarden/dept-design do you all have any good ideas on maybe a badge change we could do rather than changing the badge text? Maybe a green background shield with an inset check?

I also think it's worth a callout card explaining what the heck the check mark means. I'm thinking something similar to the autofill callout https://github.com/bitwarden/clients/pull/4738

MGibson1 commented on apps/browser/src/listeners/update-badge.ts:
```diff
@@ -130,6 +137,28 @@ export class UpdateBadge {
+      return;
+    }
+
+    try {
+      const ciphers = await this.cipherService.getAllDecryptedForUrl(tab?.url);
+      const disableBadgeCounter = await this.stateService.getDisableBadgeCounter();
+      if (ciphers.length !== 0 && !disableBadgeCounter) {
+        return;
+      }
+    } catch {
+      /** Expected CipherService "No key" error if unauthenticated. */
+    }
```
I don't see why this functionality needs to check for disabled badge counter or count urls.

MGibson1 commented on apps/browser/src/listeners/update-badge.ts:
```diff
@@ -130,6 +137,28 @@ export class UpdateBadge {
     await this.clearBadgeText();
   }
 
+  async setTrustedSite({ tab }: BadgeOptions) {
```
If this is going to replace badge text, I'd prefer to split the method into a `isTrustedVault` and incorporate the rest into the appropriate method.

Where it should eventually go depends on when we want to set the trusted site indicator. I'm thinking it makes sense in all contexts -- logged out, locked, unlocked. However, if we want it only when unlocked, that's fine too. It just changes where this is appropriate to do.

```typescript

```

MGibson1 reviewed (changes requested):
I think we need a little bit more design to decide when it's appropriate to show this checkout, then we'll know how to code this up.

danielleflinn commented:
@MGibson1 

>  I would like it to be limited to vault.bitwarden.com (or the exact vault url for self host).

This works for me

> do you all have any good ideas on maybe a badge change we could do rather than changing the badge text? Maybe a green background shield with an inset check?

From my research into this, I'm not aware of how we might change the badge to an actual icon. We can change the Extension icon (similar to the lock icon), however this is very difficult to see given we only have about 8px of space to work with without obscuring the Bitwarden logo. See [Figma archive](https://www.figma.com/file/jJd8CGRqKy98gf9kxIh4t6/SG-1038%3A-Confirm-extension-icon?node-id=36%3A105&t=lke5IoeG2hAUSyHM-1) for those iterations. Using the extension badge gives us a slightly larger area to work with that is ultimately easier to see. 

> I also think it's worth a callout card explaining what the heck the check mark means.

I like the idea of adding additional text explaining the check; what if we provide explanation text when the extension icon is hovered over? Something like "Bitwarden - on official vault.bitwarden.com site." (updating the URL if account is self hosted)

jtouchstonebw commented:
> I'm hesitant to give up my one spot for text in these badges. I do like the idea to offer some visual indication that you're on the trusted site -- though, I would like it to be limited to _vault_.bitwarden.com (or the exact vault url for self host).
> 
> @bitwarden/dept-design do you all have any good ideas on maybe a badge change we could do rather than changing the badge text? Maybe a green background shield with an inset check?
> 
> I also think it's worth a callout card explaining what the heck the check mark means. I'm thinking something similar to the autofill callout #4738

I have a few clarifying questions--

Can you clarify what you mean by green background shield with an inset check? Also when you say "badge", do you mean the box where we display the number of logins for the current URL? Or do you mean the extension icon (with the bitwarden shield logo)?

Can you elaborate more about your concerns about changing the badge text? Is it losing context into something? Are there edge cases not considered? Something else?

The suggestion about the autofill callout [#4738](https://github.com/bitwarden/clients/pull/4738) is valid and I agree we should pursue that.

MGibson1 commented:
Hi @danielleflinn and @jtouchstonebw. Sorry to the delay here. By badge I mean the entire icon (Bitwarden shield and all) -- that's actually how the lock icon works, [it's just a new image with a lock where the text can go](https://github.com/bitwarden/clients/blob/master/apps/browser/src/images/icon38_locked.png). So we can update that whole image from a code perspective pretty easily. The number is called the "badge text". Here, @willmartian is setting the text to read "✓" rather than our typical "1" or whatever for unlocked vaults.

The issue here is that we need this check mark to be present for logged out, locked, and unlocked vaults. However, for unlocked vaults, we would also like to indicate the number of vault items the user has for that site (vault.bitwarden.com, in this case).

What I was hoping for was closest to the following, but with the check box and lock icon reversed, because I want to use the lock location for text when the vault is unlocked (in chrome, ff puts text in top right).
<img width="164" alt="image" src="https://user-images.githubusercontent.com/18214891/236916326-244ae0d7-6d70-46d5-a1b6-bf0290086fe7.png">

willmartian commented:
@MGibson1 To make sure I understand: is the preference for not using `setBadgeText("✓")` only because we currently use it for the number of vault items?

If so, the [design spec](https://bitwarden.atlassian.net/browse/PM-145) for this feature has multiple stipulations for showing the check mark, one being: `The user has no matching login items OR the “Show badge counter” has been turned off in settings`

This means we won't clash with any other _current_ usage of `setBadgeText`. Is your reservation out of preserving the badge text space for future functionality?

MGibson1 commented:
ah I hadn't appreciated that. Yes I didn't want to replace the login count. if that's not happening, it seems fine

willmartian commented:
TODO: The extension icons need to be updated such that the lock is to the left of the shield (so the check badge doesn't cover the lock). 
<img width="140" alt="image" src="https://github.com/bitwarden/clients/assets/17113462/37d75f56-a2eb-4411-84e6-97d24b69a94a">

willmartian commented:
@MGibson1 I refactored this a bit and added a callout to explain what is going on! I still have an outstanding task to update the extension icons, mentioned above, but I would appreciate another review 🙏

shane-melton reviewed (approved):
Changes for Vault look good!

MGibson1 commented on apps/browser/src/images/icon38_locked.png:
⚠️ The lock looks like a different aspect ratio

MGibson1 reviewed (commented):
I know it's pedantic, but the icons matter, let's make sure these look like we want

willmartian commented:
This is old: pre-MV3 and pre-extension refresh.
